File: Phase_1/Day-5_30-07-2026/RottingOranges.java
import java.util.*;


public class RottingOranges {

    public int orangesRotting(int[][] grid) {
        int R = grid.length, C = grid[0].length, oranges = 0, ans = 0;

        Queue<int[]> q = new LinkedList<>();
        for (int i = 0; i < R; i++) {
            for (int j = 0; j < C; j++) {
                if (grid[i][j] != 0) {
                    oranges++;
                }
                if (grid[i][j] == 2) {
                    q.add(new int[]{i, j, 0});
                    grid[i][j] = 0;
                }
            }
        }

        int d[] = {0, 1, 0, -1, 0};
        while(!q.isEmpty()) {
            int[] cell = q.poll();
            int row = cell[0], col = cell[1], dy = cell[2];
            ans = dy;
            oranges--;
            for (int i = 0; i < 4; i++) {
                int ar = row + d[i], ac = col + d[i+1];
                if (ar >= 0 && ar < R && ac >= 0 && ac < C && grid[ar][ac] == 1) {
                    q.add(new int[]{ar, ac, dy+1});
                    grid[ar][ac] = 0;
                }
            }
        }
        return oranges == 0 ? ans : -1;
    }
}

File: Phase_1/Day-5_30-07-2026/RottingOranges.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

class Solution
{
public:
    const int d[5] = {0, 1, 0, -1, 0};
    int orangesRotting(vector<vector<int>> &grid)
    {
        int R = grid.size(), C = grid[0].size();
        int oranges = 0;
        queue<pair<int, int>> q;
        for (int i = 0; i < R; i++)
        {
            for (int j = 0; j < C; j++)
            {
                if (grid[i][j] != 0)
                    oranges++;
                if (grid[i][j] == 2)
                {
                    q.push(make_pair(i, j));
                    grid[i][j] = 0;
                }
            }
        }
        if (oranges == 0)
            return 0;
        int min = -1;
        while (!q.empty())
        {
            min++;
            int qsz = q.size();
            oranges -= qsz;
            while (qsz--)
            {
                auto [r, c] = q.front();
                q.pop();
                for (int i = 0; i < 4; i++)
                {
                    int ar = r + d[i], ac = c + d[i + 1];
                    if (ar >= 0 && ar < R && ac >= 0 && ac < C && grid[ar][ac] == 1)
                    {
                        q.push(make_pair(ar, ac));
                        grid[ar][ac] = 0;
                    }
                }
            }
        }
        return oranges == 0 ? min : -1;
    }
};